index/upsidedown/upsidedown_test.go

```diff
@@ -187,126 +199,119 @@ func TestIndexInsertThenDelete(t *testing.T) {
 		if err != nil {
 			t.Fatal(err)
 		}
+
+		if err := os.RemoveAll("test_dir"); err != nil {
+			t.Fatal(err)
+		}
 	}()
 
 	analysisQueue := index.NewAnalysisQueue(1)
-	idx, err := NewUpsideDownCouch(boltdb.Name, boltTestConfig, analysisQueue)
-	if err != nil {
-		t.Fatal(err)
-	}
-	err = idx.Open()
-	if err != nil {
-		t.Errorf("error opening index: %v", err)
+
+	tests := []struct {
+		desc string
+		idx  index.Index
+	}{
+		{
+			desc: "gtreap",
+			idx:  openIndex(t, gtreap.Name, gtreapTestConfig, analysisQueue),
+		},
+		{
+			desc: "boltdb",
+			idx:  openIndex(t, boltdb.Name, boltTestConfig, analysisQueue),
+		},
+		{
+			desc: "leveldb",
+			idx:  openIndex(t, goleveldb.Name, levelTestConfig, analysisQueue),
+		},
+	}
+
+	for _, test := range tests {
+		idx := test.idx
+		t.Run(test.desc, func(t *testing.T) {
+			testIndexInsertThenDelete(t, idx)
+		})
 	}
+}
+
+func testIndexInsertThenDelete(t *testing.T, idx index.Index) {
 	defer func() {
 		err := idx.Close()
 		if err != nil {
 			t.Fatal(err)
 		}
 	}()
 
-	var expectedCount uint64
-	reader, err := idx.Reader()
-	if err != nil {
-		t.Fatal(err)
-	}
-	docCount, err := reader.DocCount()
-	if err != nil {
-		t.Error(err)
-	}
-	if docCount != expectedCount {
-		t.Errorf("Expected document count to be %d got %d", expectedCount, docCount)
-	}
-	err = reader.Close()
-	if err != nil {
-		t.Fatal(err)
-	}
+	expectedCount := uint64(0)
+	validateDocCount(t, idx, expectedCount)
+	// 1 for version.
+	expectedRows := uint64(1)
+	validateRowCount(t, idx, expectedRows)
 
 	doc := document.NewDocument("1")
 	doc.AddField(document.NewTextField("name", []uint64{}, []byte("test")))
-	err = idx.Update(doc)
+	doc.AddField(document.NewTextField("name1", []uint64{}, []byte("test1")))
```

Comment:
I enhanced the test to include one index that's shared by two documents, and the other is not. I'm also happy to revert to the original test setup if you think this hurts readability.